lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -791,32 +804,44 @@ static void hoistConvertOnTopOfExt(ConvertLayoutOp convertOp) {
   if (targetType.getEncoding().isa<triton::gpu::DotOperandEncodingAttr>())
     return;
 
-  // 1. Take a backward slice of all the tensor dependencies.
-  SetVector<Value> slice;
-  DenseMap<Value, Attribute> layout;
   auto isExtOp = [](Operation *op) {
     return isa<arith::ExtSIOp, arith::ExtUIOp, arith::ExtFOp>(op);
   };
-  // Get a backward slice but don't go past ext ops
-  LogicalResult result = getConvertBackwardSlice(
-      convertOp.getOperand(), slice, targetType.getEncoding(), layout, isExtOp);
-  if (result.failed() || slice.empty())
+  // 1. Take a backward slice of all the tensor dependencies.
+  SetVector<Value> slice;
+  DenseMap<Value, Attribute> layout;
+  LogicalResult result = getRematerializableSlice(
+      convertOp.getOperand(), targetType.getEncoding(), slice, layout, isExtOp);
+  if (result.failed())
     return;
+
   Operation *extOp = nullptr;
-  // 2. Check if all the operations in the slice can be rematerialized.
-  for (Value v : slice) {
-    if (Operation *op = v.getDefiningOp()) {
-      if (!canBeRemat(op))
-        return;
-      if (isExtOp(op)) {
-        // Only apply it if there is a single ext op otherwise we would have to
-        // duplicate the convert.
-        if (extOp != nullptr)
-          return;
-        extOp = op;
+  unsigned sliceSize = slice.size();
+  for (unsigned i = 0; i < sliceSize; i++) {
+    Value v = slice[i];
+    Operation *op = v.getDefiningOp();
+    if (!op)
+      continue;
+    if (isExtOp(op)) {
+      SetVector<Value> tempSlice;
+      DenseMap<Value, Attribute> tempLayout;
+      LogicalResult result = getRematerializableSlice(
+          op->getOperand(0), layout[v], tempSlice, tempLayout);
+      // If we can rematerialize the rest of the ext slice we can ignore this
+      // ext as it won't need a convert.
+      if (result.succeeded()) {
+        slice.insert(tempSlice.begin(), tempSlice.end());
```

Comment:
Once the number of slice is changed, do you need to increase `sliceSize` as well?

Reply:
no, we only want to go through the original slice since here we get a backward slice without stopping at ext anymore. This new slice will be rematerialized without the need for convert so we don't have to consider it in looking for an ext that requires a convert.

Reply:
Got you. That makes sense